src/osd/modules/input/input_evdev.cpp

```diff
@@ -0,0 +1,745 @@
+					devices.emplace_back(&d);
+					poll_fd.emplace_back();
+					pollfd& pfd = poll_fd.back();
+					pfd.fd = d.mapped->fd;
+					pfd.events = POLLIN;
+				}
+
+			int result = poll(poll_fd.data(), poll_fd.size(), -1);
+			if(result < 0) {
+				osd_printf_warning("evdev: poll() failed: %s\n", strerror(errno));
+				continue;
+			}
+
+			if(result > 0 && (poll_fd[0].revents & POLLIN)) {
+				char octet;
+				if(read(reader_thread_pipe[0], &octet, 1)==1) {
+					if(octet == 'x')
+						break;
+					continue;
+				}
+			}
+
+			if(result != 0) {
+				input_event events[256];
+
+				for(int i=1; i<poll_fd.size(); i++) {
+					if(poll_fd[i].revents & (POLLERR|POLLHUP|POLLNVAL)) {
+						// file descriptor went away, did someone yank the device?
+						if(devices[i]->mapped)
+							devices[i]->mapped->fail();
+						continue;
+					}
+					if(poll_fd[i].revents & POLLIN) {
+						// we try to reduce the number of system calls by reading
+						// as many events as reasonable in one read() call
+						ssize_t rdlen = read(poll_fd[i].fd, events, 256*sizeof(input_event));
+						if(rdlen<0 && (errno==EWOULDBLOCK || errno==EAGAIN))
+							continue;
+						if(devices[i]->mapped) {
+							if(rdlen < sizeof(input_event))
+								devices[i]->mapped->fail();
+							else
+								devices[i]->mapped->process(events, rdlen/sizeof(input_event));
+						}
+					}
+				}
+			}
+		}
+	}
+
+	void init()
+	{
+		if(initialized)
+			return;
+		initialized = true;
+		osd_printf_verbose("evdev: Enumerating devices\n");
+		device_dir_fd = open("/dev/input/by-path", O_RDONLY|O_CLOEXEC|O_DIRECTORY);
+		if(device_dir_fd < 0) {
+			osd_printf_warning("evdev: unable to open device directory: %s\n", strerror(errno));
+			return;
+		}
+
+		DIR* idir = fdopendir(device_dir_fd);
+		while(idir) {
+			dirent* de = readdir(idir);
+			if(!de)
+				break;
+
+			char linkbuf[16];
+			ssize_t rl = readlinkat(device_dir_fd, de->d_name, linkbuf, sizeof(linkbuf)-1);
+			if(rl < 0)
+				continue;
+			linkbuf[rl] = 0;
+			if(strncmp(linkbuf, "../event", 8))
+				continue;
+
+			int ed = openat(device_dir_fd, de->d_name, O_RDONLY);
+			if(ed < 0) {
+				osd_printf_warning("evdev: unable to open device '%s', ignored: %s\n", de->d_name, strerror(errno));
+				continue;
+			}
+
+			char namebuf[64];
+			if(ioctl(ed, EVIOCGNAME(sizeof(namebuf)), namebuf)<0) {
+				// If the device doesn't report a name, use the evdev short link
+				// as its name "eventXX".  It's not _informative_ but it's
+				// distinctive enough
+				strncpy(namebuf, linkbuf+3, 63);
+				namebuf[63] = 0;
+			}
+
+			dev.emplace_back();
+			evdev_device_info& device = dev.back(); // C++17 would return back() from emplace_back()
+			
+			device.name = namebuf;
+			device.id = de->d_name;
+
+			std::fill(std::begin(device.buttons), std::end(device.buttons), 0);
+			std::fill(std::begin(device.relaxes), std::end(device.relaxes), 0);
+			std::fill(std::begin(device.absaxes), std::end(device.absaxes), 0);
```

Comment:
You know, MAME *is* C++17.  But wouldn’t it be even better to encapsulate the initialisation in the device constructor?